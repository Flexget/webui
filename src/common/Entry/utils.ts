import { RawEntry, CardType, FieldNames, MappingType, GettersType } from './types';
import { Entry, DefaultEntry } from './fields';
import {
  RawMovieEntry,
  MovieEntry,
  movieFieldList,
  MovieFields,
  MovieFieldNames,
} from './fields/movies';
import {
  RawSeriesEntry,
  SeriesEntry,
  seriesFieldList,
  SeriesFields,
  SeriesFieldNames,
} from './fields/series';

import {
  RawEpisodeEntry,
  EpisodeEntry,
  EpisodeFields,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !!entry.seriesSeason;
const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<MovieFields>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<SeriesFields>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<EpisodeFields>(EpisodeFieldNames, episodesFieldList);

const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    type: CardType.Movie,
    ...getMovieProps(entry),
  };
};

const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  type: CardType.Series,
  ...getSeriesProps(entry),
});

const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  type: CardType.Episode,
  ...getEpisodeProps(entry),
});

const toDefaultEntry = (entry: RawEntry): DefaultEntry => ({
  ...entry,
  type: CardType.Default,
});

export const toEntry = (entry: RawEntry): Entry => {
  if (isMovie(entry)) {
    return toMovieEntry(entry);
  }
  if (isSeries(entry)) {
    return toSeriesEntry(entry);
  }
  if (isEpisode(entry)) {
    return toEpisodeEntry(entry);
  }
  return toDefaultEntry(entry);
};
