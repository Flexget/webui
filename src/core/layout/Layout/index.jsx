import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Logo from 'core/layout/Logo';
import Navbar from 'core/layout/Navbar';
import SideNav from 'core/layout/SideNav';
import LoadingBar from 'common/LoadingBar';
import ErrorStatus from 'common/ErrorStatus';
import InfoStatus from 'common/InfoStatus';
import { Wrapper, Main, Header, LogoWrapper, Nav, SideBar, Content } from './styles';

class Layout extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  state = {
    sideBarOpen: (window.matchMedia && !!window.matchMedia('(min-width: 600px)').matches) || false,
  };

  toggleSideBar = () => {
    this.setState(({ sideBarOpen }) => ({
      sideBarOpen: !sideBarOpen,
    }));
  };

  render() {
    const { children } = this.props;
    const { sideBarOpen } = this.state;

    return (
      <Wrapper>
        <Header>
          <LogoWrapper>
            <Logo sideBarOpen={sideBarOpen} />
          </LogoWrapper>
          <Nav>
            <Navbar toggle={this.toggleSideBar} />
            <LoadingBar />
          </Nav>
        </Header>
        <Main>
          <SideBar>
            <SideNav sideBarOpen={sideBarOpen} toggle={this.toggleSideBar} />
          </SideBar>
          <Content open={sideBarOpen}>{children}</Content>
          <ErrorStatus />
          <InfoStatus />
        </Main>
      </Wrapper>
    );
  }
}

export default Layout;
