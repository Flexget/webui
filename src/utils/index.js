export function removeDupesIgnoreCase(values) {
  const lowerMap = Object.assign({}, ...values.map(v => ({ [v.toLowerCase()]: v })));
  const uniqueValues = new Set(values.map(v => v.toLowerCase()));
  return Array.from(uniqueValues).map(v => lowerMap[v]);
}
