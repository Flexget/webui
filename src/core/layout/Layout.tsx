import React, { useCallback } from 'react';
import ErrorStatus from 'core/status/ErrorStatus';
import InfoStatus from 'core/status/InfoStatus';
import LoadingBar from 'core/status/LoadingBar';
import { useMediaQuery, useTheme } from '@material-ui/core';
import { useOverlayState } from 'utils/hooks';
import Logo from './Logo';
import Navbar from './Navbar';
import SideNav from './SideNav';
import {
  header,
  wrapper,
  logoWrapper,
  nav,
  main,
  sidebar,
  contentWithSidebar,
  content,
} from './styles';

const Layout: React.FC = ({ children }) => {
  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.down('sm'));
  const [sidebarOpen, { toggle, close }] = useOverlayState(matches);

  const contentCssFn = useCallback(() => contentWithSidebar(sidebarOpen), [sidebarOpen]);

  return (
    <div css={wrapper}>
      <header css={header}>
        <div css={logoWrapper}>
          <Logo sidebarOpen={sidebarOpen} />
        </div>
        <nav css={nav}>
          <Navbar toggleSidebar={toggle} />
          <LoadingBar />
        </nav>
      </header>
      <main css={main}>
        <aside css={sidebar}>
          <SideNav sidebarOpen={sidebarOpen} onClose={close} />
        </aside>
        <section css={[content, contentCssFn()]}>{children}</section>
        <ErrorStatus />
        <InfoStatus />
      </main>
    </div>
  );
};

export default Layout;
