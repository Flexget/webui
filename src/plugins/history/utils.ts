import { groupBy } from 'utils/array';
import { HistoryItem, GroupByFields } from './types';

export const getGroupedHistory = (items: HistoryItem[], grouping: GroupByFields) =>
  groupBy(
    items,
    grouping,
    grouping === GroupByFields.Time ? (i: string) => i.split('T')[0] : undefined,
  );
