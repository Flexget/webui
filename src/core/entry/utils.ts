import { RawEntry, FieldNames, MappingType, GettersType, BaseEntry } from './types';
import { MovieEntry, movieFieldList, RawMovieEntry, MovieFieldNames } from './fields/movies';
import { SeriesEntry, seriesFieldList, SeriesFieldNames, RawSeriesEntry } from './fields/series';
import {
  EpisodeEntry,
  RawEpisodeEntry,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

export const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
export const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !entry.seriesSeason;
export const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry & SeriesEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode && entry.seriesName);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<RawMovieEntry>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<RawSeriesEntry>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<RawEpisodeEntry>(EpisodeFieldNames, episodesFieldList);

export const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    ...getMovieProps(entry),
  };
};

export const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  ...getSeriesProps(entry),
});

export const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  ...getEpisodeProps(entry),
});

export const toDefaultEntry = (entry: BaseEntry): BaseEntry => ({
  ...entry,
});
