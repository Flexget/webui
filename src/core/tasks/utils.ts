import { RawEntry } from 'core/entry/types';

export const toExecuteRequest = (entry?: RawEntry) => {
  if (entry) {
    const { id, title, originalUrl, ...fields } = entry;
    return { title, url: originalUrl, fields };
  }
  return { title: '', url: '', fields: {} };
};
