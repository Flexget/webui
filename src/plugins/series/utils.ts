import { RawEntry } from 'core/entry/types';
import { Show, Episode } from './types';

export const showToEntry = (show: Show): RawEntry => ({
  title: show.name,
  originalUrl: '',
  quality: 'unknown',
  seriesName: show.name,
});

export const episodeToEntry = (show: Show, episode: Episode): RawEntry => ({
  title: episode.identifier,
  originalUrl: '',
  quality: 'unknown',
  seriesName: show.name,
  seriesSeason: episode.season,
  seriesEpisode: episode.number,
  seriesId: episode.identifier,
});
