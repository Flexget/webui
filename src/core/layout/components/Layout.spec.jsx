import React from 'react';
import renderer from 'react-test-renderer';
import Layout from 'core/layout/components';
import { themed, router, provider } from 'utils/tests';
import fetchMock from 'fetch-mock';

function renderLayout() {
  return provider(router(themed(
    <Layout>
      <div />
    </Layout>
  )), { router: { location: { } }, version: {}, status: { loading: {} }, routes: {} });
}
describe('common/layout', () => {
  beforeEach(() => {
    fetchMock
      .get('/api/server/version', {});
  });

  it('renders correctly', () => {
    const tree = renderer.create(renderLayout()).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
