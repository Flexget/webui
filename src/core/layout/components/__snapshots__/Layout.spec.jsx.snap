// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/layout renders correctly 1`] = `
.glamor-21 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  min-height: 100vh;
}

.glamor-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 5rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  z-index: 2;
  position: fixed;
  width: 100%;
}

.glamor-1 {
  height: 5rem;
}

.glamor-0 {
  background: #263238 url(test-file-stub) no-repeat center;
  background-size: 17.5rem;
  height: 100%;
  -webkit-transition: width 300ms cubic-bezier(0.4,0,0.2,1) 0ms,background-size 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: width 300ms cubic-bezier(0.4,0,0.2,1) 0ms,background-size 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.glamor-9 {
  height: 5rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.glamor-8 {
  position: static;
}

.glamor-7 {
  background-color: #ef6c00;
  color: #fff;
  min-height: 5rem;
}

.glamor-2 {
  color: #fff;
}

.glamor-3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.glamor-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-top: 10rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100vh;
}

.glamor-18 {
  overflow-y: auto;
}

.glamor-17 {
  position: relative;
  background-color: #263238;
  height: calc(100vh - 10rem);
  width: 0;
  -webkit-transition: width 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: width 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  border-right: none !important;
}

.glamor-16 {
  width: inherit;
  height: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.glamor-11 {
  width: inherit;
}

.glamor-15 {
  display: none;
}

.glamor-12 {
  margin: 0;
}

.glamor-19 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-y: auto;
  padding: 1rem;
  opacity: 1;
  -webkit-transition: opacity 300ms cubic-bezier(0.4,0,0.2,1) 0ms,margin-left 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 300ms cubic-bezier(0.4,0,0.2,1) 0ms,margin-left 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

@media (min-width:600px) {
  .glamor-10 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:600px) {
  .glamor-0 {
    width: 6rem;
    background-size: 25rem;
  }
}

@media (min-width:600px) {
  .glamor-20 {
    padding-top: 5rem;
  }
}

@media (min-width:600px) {
  .glamor-17 {
    height: calc(100vh - 5rem);
    width: 6rem;
  }
}

@media (max-width:959.95px) {
  .glamor-19 {
    margin-left: 0;
    opacity: 1;
    display: block;
  }
}

<div
  className="glamor-21 e1ovaqso0"
>
  <header
    className="glamor-10 e1ovaqso2"
  >
    <div
      className="glamor-1 e1ovaqso3"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <div
          className="glamor-0 ee44yi0"
          open={false}
        />
      </a>
    </div>
    <nav
      className="glamor-9 e1ovaqso4"
    >
      <header
        className="MuiPaper-root-9 MuiPaper-shadow4-15 MuiAppBar-root-1 MuiAppBar-positionFixed-2 MuiAppBar-colorPrimary-7 mui-fixed glamor-8 e1pzz03b0"
      >
        <div
          className="MuiToolbar-root-36 MuiToolbar-gutters-37 glamor-7 e1pzz03b1"
        >
          <button
            className="MuiButtonBase-root-44 MuiIconButton-root-38 glamor-2 e1pzz03b2"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label-43"
            />
            <span
              className="MuiTouchRipple-root-46"
            />
          </button>
          <h2
            className="MuiTypography-root-53 MuiTypography-title-59 MuiTypography-colorInherit-72"
          />
          <div
            className="glamor-3 e1vib4761"
          />
          <a
            href="/config"
            onClick={[Function]}
          >
            <button
              aria-label="Config"
              className="MuiButtonBase-root-44 MuiIconButton-root-38 glamor-2 e1pzz03b2"
              disabled={false}
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label-43"
              />
              <span
                className="MuiTouchRipple-root-46"
              />
            </button>
          </a>
          <a
            href="/log"
            onClick={[Function]}
          >
            <button
              aria-label="Log"
              className="MuiButtonBase-root-44 MuiIconButton-root-38 glamor-2 e1pzz03b2"
              disabled={false}
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label-43"
              />
              <span
                className="MuiTouchRipple-root-46"
              />
            </button>
          </a>
          <button
            aria-label="Manage"
            className="MuiButtonBase-root-44 MuiIconButton-root-38 glamor-2 e1pzz03b2"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label-43"
            />
            <span
              className="MuiTouchRipple-root-46"
            />
          </button>
        </div>
      </header>
    </nav>
  </header>
  <main
    className="glamor-20 e1ovaqso1"
  >
    <aside
      className="glamor-18 e1ovaqso5"
    >
      <div
        className="MuiDrawer-docked-88"
      >
        <div
          className="MuiPaper-root-9 MuiPaper-shadow0-11 MuiDrawer-paper-89 glamor-17 MuiDrawer-paperAnchorLeft-90 MuiDrawer-paperAnchorDockedLeft-94"
        >
          <div
            className="glamor-16 egdat1"
          >
            <ul
              className="MuiList-root-99 MuiList-padding-100 glamor-11 egdat3"
            />
            <div
              className="egdat2 glamor-15 ekjxbjj0"
            >
              <p
                className="glamor-12 ekjxbjj1"
              >
                Version Info
              </p>
              <p
                className="glamor-12 ekjxbjj1"
              >
                Flexget: 
                 
              </p>
              <p
                className="glamor-12 ekjxbjj1"
              >
                API: 
              </p>
            </div>
          </div>
        </div>
      </div>
    </aside>
    <section
      className="glamor-19 e1ovaqso6"
      open={false}
    >
      <div />
    </section>
  </main>
</div>
`;
